guides/common/modules/proc_mqtt-tuning.adoc

```diff
@@ -0,0 +1,28 @@
+[id="REX_pull_provider_{context}"]
+= REX pull provider 
+
+Remote Execution has been around for a while now and has served SSH-based and Ansible execution very well. In many environments these are not options due to either security or networking concerns.
+
+Katello-agent has provided an alternative for a long time that worked in an environment where SSH was not available or the Foreman/Katello server or smart proxies did not have the ability to connect to a client. However katello-agent only supported a small set of actions around updating packages. There has been a long desire to create a ‘pull’ based provider that has the features of REX but with a deployment model similar tokatello-agent. 
+
+The solution uses MQTT as its messaging protocol and includes an MQTT client running on each Host. In orderto tune the MQTT, you can follow the following procedure on the Satellite or capsule server:
+
+.*Procedure*
+. Enable the feature using: `satellite-installer --foreman-proxy-plugin-remote-execution-script-mode=pull-mqtt` (this will change the ReX transport mode for the Satellite or capsule from the default SSH one to MQTT). It’s important to mention that one Satellite/capsule server can only use one transport mode: SSH or MQTT
+
+. In order to bump the default number of clients, create the following system unit definition file and set its contents as specified in the below given example (this will allow the `mosquitto` server to handle ~5000 clients):
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+cat >/etc/systemd/system/mosquitto.service.d/limits.conf <<__EOF__
+[Service]
+LimitNOFILE=5000
```

Comment:
If this is common (and I'm quite sure it is), I'd prefer to support this properly in our installer. I'm worried our installer actually purges this file again.

Having said that: EL8 has a default file limit of 262144 so you're SERIOUSLY lowering it. See `systemctl show -p DefaultLimitNOFILE` and `systemctl show -p LimitNOFILE foreman-mosquitto.service` as well.

Are you sure you weren't testing this on EL7 which has a much lower limit of 4096?

Reply:
Hello @ekohl , we are already calling it here [1] right? 

[1] https://github.com/theforeman/foreman-documentation/blob/master/guides/common/modules/con_performance-tuning-quickstart.adoc. 

@pablomh could you please answer about the 4096 limit thing ?

Reply:
What I mean is that in our documentation we want to avoid users managing individual files. If this is something that we want to support, the installer should have some option to do so. It may take tuning profiles into account, depending on how we choose to implement it.

Reply:
Hi @ekohl. In our `mosquitto` testing we found the following [issue] in a `medium` tuned (not that it matters in this case) capsule when trying to register the 1015th client:
~~~
Unable to accept new connection, system socket count has been exceeded. Try increasing "ulimit -n" or equivalent.
~~~
We were able to overcome the issue making the modifications that we propose in this PR, so maybe the `mosquitto` system unit is not inheriting properly that property.

[issue]: https://bugzilla.redhat.com/show_bug.cgi?id=2132774

Reply:
https://github.com/systemd/systemd/blob/23f3a6f5ff864fd26063c6c35fdaa6d85de566c7/NEWS#L5880-L5914 does have quite a bit to say on this. 

> Note that the soft limit remains at 1024 for compatibility reasons: the
          traditional UNIX select() call cannot deal with file descriptors >=
          1024 and increasing the soft limit globally might thus result in
          programs unexpectedly allocating a high file descriptor and thus
          failing abnormally when attempting to use it with select() (of
          course, programs shouldn't use select() anymore, and prefer
          poll()/epoll, but the call unfortunately remains undeservedly popular
          at this time).

Then later:

> Programs that want to take benefit of the increased limit have to "opt-in" into
          high file descriptors explicitly by raising their soft limit.

That 1015 is seriously close to 1024 that it makes me suspect it's observing the soft limit by default and I suspect that if Mosquitto was modified to raise its soft limit, it would automatically get a lot more file descriptors.

I think we should pursue the path of fixing it upstream first. For now I've opened https://github.com/eclipse/mosquitto/issues/2674 to see if it's a viable solution.

Reply:
Thank you Ewoud! Do we want to add a reference to that upstream ticket to the Tuning Guide in this PR?